tracing/tests/max_level_hint.rs

```diff
@@ -0,0 +1,31 @@
+mod support;
+
+use self::support::*;
+use tracing::Level;
+
+#[test]
+fn max_level_hints() {
+    let (subscriber, handle) = subscriber::mock()
+        .with_max_level_hint(Level::INFO)
+        .with_filter(|meta| {
+            assert!(
+                dbg!(meta).level() <= &Level::INFO,
+                "a TRACE or DEBUG event was dynamically filtered: "
+            );
+            true
+        })
+        .event(event::mock().at_level(Level::INFO))
+        .event(event::mock().at_level(Level::WARN))
+        .event(event::mock().at_level(Level::ERROR))
+        .done()
+        .run_with_handle();
+
+    tracing::subscriber::set_global_default(subscriber).unwrap();
+
+    tracing::info!("doing a thing that you might care about");
+    tracing::debug!("charging turboencabulator with interocitor");
+    tracing::warn!("extremely serious warning, pay attention");
+    tracing::trace!("interocitor charge level is 10%");
```

Comment:
I'm assuming this trace event doesn't trigger the assertion above but I'm not sure why.

Reply:
It doesn't trigger the assertion because it's disabled by the global max level hint provided by the subscriber. Because the subscriber has told us that it will _never_ enable anything above `INFO`, we can skip `TRACE` or `DEBUG` events without calling `enabled`. The assertion here is testing that this actually works.

Reply:
I added comments to the tests that should explain what they're testing for and what those assertions mean: cabca0d77f93810672c98f91b2b70086d2e8ef61. Does that help?

Reply:
yup! that makes perfect sense